### PR TITLE
Implement and/or filters on subscriptions

### PR DESCRIPTION
**Pull request by a-alle**
# Description

Allows for combining filters using AND/ OR for subscriptions.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [ ] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [ ] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at
 http://neo4j-graphql-toolbox-stage-2099.surge.sh (=> for local development; use for instance [Neo4j Desktop](https://neo4j.com/download/))
or
 https://neo4j-graphql-toolbox-stage-2099.surge.sh (=> use to connect to an [Aura Free](https://neo4j.com/cloud/aura-free/) dbms)

File: .changeset/proud-houses-love.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@neo4j/graphql": minor
+---
+
+Allows for combining filters with AND/OR when subscribing to events.
```
Comment (angrykoala) [inline .changeset/proud-houses-love.md:5]: I'm not 100% sure, but I think we can remove the `for`

```suggestion
Allows combining filters with AND/OR when subscribing to events.
```

File: packages/graphql/src/schema/resolvers/subscriptions/update-diff-filter.ts [hunk 2]
```diff
@@ -19,16 +19,19 @@
...
 export function updateDiffFilter(event: SubscriptionsEvent): boolean {
     if (event.event !== "update") {
         return true;
     }
-
-    const sameLength = Object.keys(event.properties.old).length === Object.keys(event.properties.new).length;
-    if (!sameLength) return true;
-    const sameProperties = compareProperties(event.properties.old, event.properties.new);
-    if (!sameProperties) return true;
+    if (!haveSameLength(event.properties.old, event.properties.new)) {
```
Comment (angrykoala) [inline packages/graphql/src/schema/resolvers/subscriptions/update-diff-filter.ts:28]: :+1: Much more readable

File: packages/graphql/tests/e2e/subscriptions/filtering/create.e2e.test.ts [hunk 3]
```diff
@@ -147,6 +147,232 @@ describe("Create Subscription with optional filters valid for all types", () =>
         expect(wsClient.events).toEqual([]);
     });
 
+    test("create subscription with where OR", async () => {
```
Comment (angrykoala) [inline packages/graphql/tests/e2e/subscriptions/filtering/create.e2e.test.ts:150]: I may have missed some test cases, but I would like to make sure the following cases are tested:
```
(where: { OR: [{ title: "movie1"}] })    // A single element inside OR
(where: {OR: [{ title: "movie1"}, AND: [{title: "movie2" } ,  {title: "Movie3"}] // Nested conditionals

```

File: packages/graphql/src/schema/resolvers/subscriptions/utils/compare-properties.ts [hunk 4]
```diff
@@ -154,19 +151,51 @@ function parseFilterProperty(key: string): { fieldName: string; operator: string
...
-        const receivedValue = receivedProperties[fieldName];
-        const fieldMeta = node.primitiveFields.find((f) => f.fieldName === fieldName);
-        const checkFilterPasses = getFilteringFn(operator);
-        if (!checkFilterPasses(receivedValue, v, fieldMeta)) {
-            return false;
+        if (Object.keys(multipleConditionsAggregationMap).includes(k)) {
+            const comparisonResultsAggregationFn = multipleConditionsAggregationMap[k];
+            if (!comparisonResultsAggregationFn) {
+                return false;
+            }
```
Comment (angrykoala) [inline packages/graphql/src/schema/resolvers/subscriptions/utils/compare-properties.ts:184]: Could this happen? or is it just typescript complaining?

**Review by angrykoala — CHANGES_REQUESTED** [review]
Looking good, just some minor comments and a few test cases that we should make sure are there

File: packages/graphql/src/schema/resolvers/subscriptions/utils/compare-properties.ts (hunk 4 above)
Comment (a-alle) [reply packages/graphql/src/schema/resolvers/subscriptions/utils/compare-properties.ts:184]: cannot happen, I used to check with hardcoded "AND", "OR" on line 180 but then changed to Object.keys. Will remove this check :D

File: packages/graphql/tests/e2e/subscriptions/filtering/create.e2e.test.ts (hunk 3 above)
Comment (a-alle) [reply packages/graphql/tests/e2e/subscriptions/filtering/create.e2e.test.ts:150]: on it

Comment (a-alle) [reply packages/graphql/tests/e2e/subscriptions/filtering/create.e2e.test.ts:150]: done

File: docs/modules/ROOT/pages/subscriptions/filtering.adoc [hunk 5]
```diff
@@ -144,3 +149,32 @@ movieDeleted(where: {genre_NOT: "Comedy"}) {
...
+                {releasedIn_GT: 2000, averageRating_GT: 7}
+            ]}
+        ]}
+    ]
+}) {
+    updatedMovie {
+        title
+    }
+}
+----
```
Comment (angrykoala) [inline docs/modules/ROOT/pages/subscriptions/filtering.adoc:180]: We could consider just using a link here to the top level filtering page and mention that AND/OR filters are supported

**Review by angrykoala — APPROVED** [review]

**Pull request merged**